File: eco-core/core-backend/src/main/java/com/willfp/eco/internal/EcoCleaner.java
package com.willfp.eco.internal;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.proxy.Cleaner;
import com.willfp.eco.internal.proxy.EcoProxyFactory;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URLClassLoader;

public class EcoCleaner implements Cleaner {
    @Override
    public void clean(@NotNull final EcoPlugin plugin) {
        if (!plugin.getProxyPackage().equalsIgnoreCase("")) {
            EcoProxyFactory factory = (EcoProxyFactory) plugin.getProxyFactory();
            factory.clean();
        }

        Plugins.LOADED_ECO_PLUGINS.remove(plugin.getName().toLowerCase());

        if (plugin.getClass().getClassLoader() instanceof URLClassLoader urlClassLoader) {
            try {
                urlClassLoader.close();
            } catch (IOException e) {
                // Do nothing.
            }
        }

        System.gc();
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/EcoCleaner.kt
package com.willfp.eco.internal

import com.willfp.eco.core.EcoPlugin
import com.willfp.eco.core.items.Items
import com.willfp.eco.core.proxy.Cleaner
import com.willfp.eco.internal.proxy.EcoProxyFactory
import java.net.URLClassLoader

class EcoCleaner: Cleaner {
    override fun clean(plugin: EcoPlugin) {
        if (plugin.proxyPackage.isNotEmpty()) {
            val factory = plugin.proxyFactory as EcoProxyFactory
            factory.clean()
        }

        Plugins.LOADED_ECO_PLUGINS.remove(plugin.name.lowercase())

        for (customItem in Items.getCustomItems()) {
            if (customItem.key.namespace.equals(plugin.name.lowercase(), ignoreCase = true)) {
                Items.removeCustomItem(customItem.key)
            }
        }

        val classLoader = plugin::class.java.classLoader

        if (classLoader is URLClassLoader) {
            classLoader.close()
        }

        System.gc()
    }
}